src/journal/journal-file.c

```diff
@@ -372,6 +407,13 @@ JournalFile* journal_file_close(JournalFile *f) {
         if (f->mmap && f->cache_fd)
                 mmap_cache_free_fd(f->mmap, f->cache_fd);
 
+        if (f->fd >= 0 && f->truncate_on_close) {
+                uint64_t p;
+
+                if (!journal_file_tail_end(f, &p))
+                        (void) ftruncate(f->fd, p);
+        }
```

Comment:
I'd add some debug logging here in case of failure:

```c
r = journal_file_tail_end(f, &p);
if (r < 0)
        log_debug_errno(r, "Failed to determine file contents size, ignoring: %m");
else {
        if (ftruncate(f->f, p) < 0)
                log_debug_error(errno, "Failed to truncate file to contents size, ignoring: %m");
}
```